keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,970 @@
+I created PDB for this deployment and set `.spec.minAvailable` to 3. I wanted the PDB to be 
+guaranteed even in case of scheduler preemption.
+
+#### Story 2
+I created a Tensorflow distributed job that requires a minimum of 5 workers running.
+I created PDB for the job and set `.spec.minAvailable` to 5. I wanted the PDB to be 
+guaranteed even in case of scheduler preemption to ensure the stability of the whole job.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+If a user sets a PodDisruptionBudget (PDB) for some low-priority pods and sets the `DisruptionPolicy` 
+to `PodDisruptionBudgetStrict`, high-priority(except `system-cluster-critical` or `system-node-critical`) 
+pods will not be able to violate the PDB and preempt these pods during the scheduling process. 
+This may result in high-priority(except `system-cluster-critical` or `system-node-critical`) pods 
+being unable to schedule while low-priority pods continue to run normally. 
+
+Although the above situations may arise, due to the fact that PriorityClass is created and managed by 
+cluster administrators with no permission for application owners to perform actions, administrators 
+are able to uniformly configure according to the requirements. Additionally, implementation will 
+include the addition of additional logging or event descriptions to clearly inform the user of the 
+reason why preemption did not occur.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In order to address the issue mentioned above, a new field `DisruptionPolicy` will 
+be added to `PriorityClass`. Users will be able to set this field to 
+determine whether or not PDB violation will be allowed during scheduler preemption. The 
+default value for this field will be `PodDisruptionBudgetBestEffort`, which is 
+consistent with current behavior.  Additionally, in order to avoid system stability 
+issues, even if the victim's PDB is not allowed to be violated during the preemption 
+process, the preempter with priority ClassName of `system-cluster-critical` or `system-node-critical` 
```

Comment:
I wonder if the system classes shouldn't be enforced by the API? Maybe by having a `SystemPriority` preemption policy that would be used only for these and would override the `DisruptionPolicy`?

I know, it seems like we are running in a circle about who has the most rights. So it is question if we want to have this hard coded or customizable?

Reply:
But that's pretty much the semantics of the field `PreemptionPolicy`.

I think the point of this KEP is that giving all the control to the higher priority class is not enough for some users.

Reply:
I think we need to reconsider having `PreemptLowerPriorityWithoutViolatePDB` in the higher priority pod.

And if a pod is not important enough to block preemption, it probably shouldn't have a PDB.

Reply:
>  I think we need to reconsider having PreemptLowerPriorityWithoutViolatePDB in the higher priority pod.
> And if a pod is not important enough to block preemption, it probably shouldn't have a PDB.

It's true that what you said makes sense, but this may require the user to configure strictly in this way, which would result in a high cost for the user. For example, he needs to check the PDB of all pods with Priority Class. In this way, it might be better to configure only the Priority Class that he needs to protect, with a lower cost.

Reply:
Please clarify the semantics of your proposal when the user sets both PreemptionPolicy and DisruptionPolicy, in other words, the precedence.

Reply:
PriorityClass used to cover the semantics of "preemptor" only, and now it starts to take "preemptee" into consideration. So +1 to define the precedence.

In a real-world case, disruption is more unacceptable than unscheduled. So I'd suggest always prioritizing DisruptionPolicy over PreemptionPolicy. This can also guide the users to opt-in (hard) disruption policy for their workloads to avoid disruption.

> I wonder if the system classes shouldn't be enforced by the API? Maybe by having a SystemPriority

I'd like to make the semantics explicitly defined in API as well, but to ease understanding the semantics, IMO we should apply on the "preemptee" side:

- PDBStrict (means "once scheduled, nobody can preempt me")
- PDBStrictExceptForSystemPods (means "once scheduled, only high-priority pods that are system pods can preempt me")

> I think we need to reconsider having `PreemptLowerPriorityWithoutViolatePDB` in the higher priority pod.

I'd hold off adding more policies to the "preemptor" side, as it will make the semantics of disruption/preemption ambiguous. If workloads don't want to be preempted, they should declare DisruptionPolicy explicitly; instead of expecting the preemptors to have "mercy" on them.

Reply:
what you are calling "mercy" could be a compromise to the stability of the system. Higher priority pods might have higher priority because they are critical to the system. How do we get the power back to those pods?

Reply:
> How do we get the power back to those pods?

(Not quite sure I understand the power correctly) I mean with the introduction of disruption policy, the semantics is already not that intuitive, so I'd suggest keeping the **preemptor** side simple, unless it's a must-need. Otherwise, we may end up with a super completed semantics of preemption. Suppose we introduce other hard disruption policies, like `ABCStrict`, `XYZStrict`, and we want to follow the same pattern of "PreemptLowerPriorityWithoutViolatePDB", we will end up with introducing more preemption policies like `PreemptLowerPriorityWithoutViolateABC`, `PreemptLowerPriorityWithoutViolateXYZ` and `PreemptLowerPriorityWithoutViolateABCAndXYZ` - this is way too complicated.

So if we set a tone on day 1 to accommodate preemption customization on "preemptee" side _only_, it may make the semantics more maintainable, extensible, and easily understood.

Reply:
> But we already have policies from the preemptor side.

yes, but those are pretty fundamental ones (PreemptLower and PreemptNever).

> So we keep all configurations in the preemptor side, we are already minimizing confusion.

But we're adding semantics (disruption policy) on preemptee side. I doubt if we can afford more complex policy combinations on **_both_** sides.

Reply:
> Please clarify the semantics of your proposal when the user sets both PreemptionPolicy and DisruptionPolicy, in other words, the precedence.

Added in the design: The PreemptionPolicy is used to describe the behavior of the pods associated with the PriorityClass during preemption as the preemptor. The DisruptionPolicy is used to describe the policy of the pods associated with the PriorityClass during preemption as the victim. During a preemption process, if the preemptor is configured with the PreemptionPolicy and the victim is configured with the DisruptionPolicy, the DisruptionPolicy takes priority over the PreemptionPolicy.

Reply:
> But we already have policies from the preemptor side. So we keep all configurations in the preemptor side, we are already minimizing confusion.

I still feel that this approach would result in a high cost. Currently, we cannot guarantee that all users will use PDB effectively. That is, if a pod is not important enough to block preemption, it probably shouldn't have a PDB. 

In this sense, for users who need to use our feature, if they want to protect a certain service, they need to set all other PriorityClasses to Strict, whereas only setting their own service's corresponding PriorityClass to Strict would be easier.

Reply:
It is easier, yes, but that's not what I'm debating here. My points are:
- It adds complexity when used in combination with PreemptionPolicy (thanks for clarifying the precedence, though)
- It gives too much power to low priority classes. How can a higher priority class take the power back? Consider a case where a service is more important than a Pod with PDB and DisruptionPolicy, but not important enough to be system-critical.

Reply:
> It gives too much power to low priority classes

Yes it is, so I do appreciate the discussion earlier that it's agreed this disruption policy should be imposed to PrirotyClass (cluster-scoped) instead of PDB (namespaced). Cluster SRE/Admin should be responsible for PriorityClass's management, and when they opt-in this feature, they should know the implication of "XYZStrict".

> How can a higher priority class take the power back?

We define clearly define it in disruption policy. For example:

```yaml
disruptionPolicy:
  XYZStrictExcluding: # means no PC can disrupt me except for the followings
  - system-cluster-critical
  - system-node-critical
  - a-particular-high-priority
```

Reply:
Maybe it should be a numeric value. Every pod with priority higher than the given value is allowed to preempt the Workload. And that value can't be higher than the value for system-node-critical.

It's a bit more resilient to changes to the set of PriorityClasses.

Reply:
> Maybe it should be a numeric value.

Makes sense. The spec can be like:

```yaml
disruptionPolicy:
- name: XYZStrict
  excludePriorityValueGT: <numeric value>
```

or, more expressible

```yaml
disruptionPolicy:
- name: XYZStrict
  priorityClassExclusionSelector:
  ...
```

Reply:
```yaml
allowPreemptionByPriorityGreaterThan: <value>
```

When it's not set, it's equal to the priority of the class  +1, which gives the current behavior.

Reply:
```yaml
disruptionPolicy:
- name: PDBStrict
   allowPreemptionByPriorityGreaterThan: <numeric value>
```
+1 it's more resilient

Reply:
I don't like how the proposal adds "strict" and then the next field is an exception. And in combination with preemptionPolicy it becomes a nightmare to explain which field takes precedence.

Reply:
And why does it need to be an array?

Reply:
> I don't like how the proposal adds "strict" and then the next field is an exception. 

How about this one? @alculquicondor  It will be easier and clearer.

```go
type PriorityClass struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Value int32
	GlobalDefault bool
	Description string
	PreemptionPolicy *core.PreemptionPolicy
        
        // If not set, it equals to the `Value + 1`
	AllowDisruptionByPriorityGreaterThan int32
}
```

> And why does it need to be an array?

Consider if there will be other policies, but there shouldn't be.

Reply:
> in combination with preemptionPolicy it becomes a nightmare to explain which field takes precedence.
> And why does it need to be an array?

It's the same question indeed. If there is no need to have multiple disruption policies, I agree flattening them is more clear:

```
disruptionPolicy: PDBStrict
allowPreemptionByPriorityGreaterThan: <value>
```

Reply:
How about just use `AllowDisruptionByPriorityGreaterThan` directly ?  
```yaml
allowDisruptionByPriorityGreaterThan: <value>
```

Greater than Value, less than or equal to AllowDisruptionByPriorityGreaterThan is Strict. 
Greater than AllowDisruptionByPriorityGreaterThan is BestEffort.

Reply:
Maybe `GreaterThanOrEqual` would be better so that you can use node-system-critical easily instead of users having to calculate node-system-critical-1.

Reply:
> How about just use AllowDisruptionByPriorityGreaterThan directly?

Do you mean with a single field and hardcode underlying scop to PDB? or two fields?

Reply:
Updated by the suggestion from https://github.com/kubernetes/enhancements/pull/3755#discussion_r1094884915 
Pls take a look. @alculquicondor @Huang-Wei

Reply:
also /cc @liggitt

Reply:
I wouldn't try to design the ideal API at the KEP time, we can always update the KEP based on the api review.

Reply:
+1 to leave API spec to API review, but I do that having a struct `DisruptionPolicy` (and under that define a "PDB" _explicitly_) will be more extensible.